### PR TITLE
Virtual Domains now have certain areas protected from ghost role interference

### PR DESCRIPTION
**Pull request by Rhials**
## About The Pull Request

This touches up on the bitrunning ghost roles that come with some maps, namely Corsair Cove and Syndicate Assault.

The gist of it is: Ghost role spawners and Digital Anomalies (the random event boss mobs) are now restricted from entering the VDOM safehouse, and other areas where critical equipment is stored.

Here's an example from Syndicate Assault -- The X-ed out area is considered "out of bounds" for digital anomalies/ghost roles:
image

Additionally, this also fixes the matter of pirate ghost role spawns creating their own antag datum/pirate team, which would carry into the roundend report. Since these are no longer legitimate pirate spawners and are now specifically designed spawners for virtual domains.

Naturally, emagging the server jailbreaks all of these restrictions and notifies any virtual entities.

The new subtype of spawners should also be scalable enough that new VDOMs should be able to implement new ghost role spawners with ease.
## Why It's Good For The Game

It's one thing to have sentient mobs to fight, which can shake up the otherwise somewhat static nature of bitrunning maps, but when players are tossing equipment, spawncamping, or otherwise making it impossible for the runners to fight them it ends up being unfun for everyone involved. You can't get into a good fight with a bitrunner avatar if their only recourse is to wipe the map and everything (YOU) in it.

This ensures a level of fairness between the (typically vindictive) ghost roles of a VDOM and the players.

Also, pirate spawns don't make a new pirate team/datum. That's one of the fixes I was aiming for with this.
## Changelog
:cl: Rhials
balance: Virtual domain ghost roles can no longer enter the safehouse/"equipment" areas of a domain.
fix: Pirate virtual domain ghost roles will no longer make a pirate team antag datum.
/:cl:

### CODE REVIEW

**Comment by Rhials** [comment]
This should also close #80167, as the virtual syndicate outfit isn't given a headset (although I think this was solved by another PR already?)

File: code/datums/components/virtual_entity.dm [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/datum/component/virtual_entity
+	///The cooldown for balloon alerts, so the player isn't spammed while trying to enter a restricted area.
+	COOLDOWN_DECLARE(OOB_cooldown)
+
+/datum/component/virtual_entity/Initialize()
+	. = ..()
+	RegisterSignal(parent, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(on_parent_pre_move))
+
+/datum/component/virtual_entity/proc/on_parent_pre_move(atom/movable/source, atom/new_location)
+	var/area/location_area = get_area(new_location)
```
Comment (MrMelbert) [inline code/datums/components/virtual_entity.dm:11]:
```suggestion
/datum/component/virtual_entity/proc/on_parent_pre_move(atom/movable/source, atom/new_location)
	SIGNAL_HANDLER
	var/area/location_area = get_area(new_location)
```

File: code/__DEFINES/_flags.dm [hunk 2]
```diff
@@ -129,6 +129,8 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define EVENT_PROTECTED (1<<15)
 /// This Area Doesn't have Flood or Bomb Admin Messages, but will still log
 #define QUIET_LOGS (1<<16)
+/// This area does not allow virutal entities to enter.
```
Comment (jlsnow301) [inline code/__DEFINES/_flags.dm:132]:
```suggestion
/// This area does not allow virtual entities to enter.
```

File: code/datums/components/virtual_entity.dm (hunk 1 above)
Comment (jlsnow301) [reply code/datums/components/virtual_entity.dm:11]: also write an autodoc comment for this pls- just what its doing/when it fires

**Comment by Rhials** [comment]
Small problem, this interferes with emagged servers/digital anomalies leaving the VDOM. I'll cook up a solution in a bit but it should be good to merge after.

**Comment by jlsnow301** [comment]
> Small problem, this interferes with emagged servers/digital anomalies leaving the VDOM. I'll cook up a solution in a bit but it should be good to merge after.

Good thinking. Imo this might be a good time to give screen alerts to the virtual entities that the server has been emagged. Delete the component thereafter?

**Comment by Rhials** [comment]
Should be ready now. I implemented a to_chat but not a screen alert, which gets sent to entities both when the server is emagged and when they spawn after the server has been emagged.

Also, completely unrelated to what was worked on here, I discovered #82973. Human mobs really don't like leaving the simulation I guess.

File: code/datums/components/virtual_entity.dm [hunk 3]
```diff
@@ -5,7 +5,14 @@
...
 /datum/component/virtual_entity/Initialize()
 	. = ..()
+
+	for(var/obj/machinery/quantum_server/server in SSmachines.get_machines_by_type(/obj/machinery/quantum_server))
+		if(server.obj_flags & EMAGGED)
+			jailbreak_mobs() //This just sends a message and self-deletes, a bit messy but it works.
+			return
+
 	RegisterSignal(parent, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(on_parent_pre_move))
+	RegisterSignal(parent, COMSIG_BITRUNNER_SERVER_EMAGGED, PROC_REF(jailbreak_mobs))
```
Comment (jlsnow301) [inline code/datums/components/virtual_entity.dm:15]: does parent receive this signal or ssdcs? if we had a ref to the server we could conditionally register it if it wasn't already emagged

Comment (Rhials) [reply code/datums/components/virtual_entity.dm:15]: It gets sent as a global signal. I'm assuming that means ssdcs get sent this signal here.

Comment (jlsnow301) [reply code/datums/components/virtual_entity.dm:15]: have you tested it

File: code/datums/components/virtual_entity.dm [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/// Handles all special considerations for "virtual entities" such as bitrunning ghost roles or digital anomaly antagonists.
```
Comment (jlsnow301) [inline code/datums/components/virtual_entity.dm:1]: there's a component folder just for bitrunning related stuff in the module :(

**Comment by Admiral-Gigaton** [comment]
> Small problem, this interferes with emagged servers/digital anomalies leaving the VDOM. I'll cook up a solution in a bit but it should be good to merge after.

wait what, they can leave?? wtf does the emagged server even do? i didnt know you could emag them

**Comment by jlsnow301** [comment]
> wait what, they can leave?? wtf does the emagged server even do? i didnt know you could emag them

yes, you can see when a server is emagged by its blinking red light

the idea is that bitrunners are weak and have to call upon their digital friends to do their dirty work, which are unloyal and dangerous, possibly killing them in the process

File: code/datums/components/virtual_entity.dm (hunk 3 above)
Comment (Rhials) [reply code/datums/components/virtual_entity.dm]: okay you were right on this, parent was DCS idk what I was thinking

**Comment by Rhials** [comment]
New solution works and routes through on_threat_created. I'm leaving this as a draft for the time being because I would like to add a Digital Anomaly antag datum subtype so ghost roles aren't lumped in with the "alive" crew tab.

**Review by Maurukas — APPROVED** [review]
Maps look good, so do the areas and what I understand of the components.

**Pull request merged**